[KEYCLOAK-5072] - Improve performance of JSPolicyProvider

janslow opened the pull request:
Creating a new `NashornScriptEngine` is quite slow, so only create a single one and just use new Bindings per evaluation (which are much cheaper). As the engine/global scopes are empty and we're creating a new scope per evaluation, this will still keep each evaluation independent of one another.

In addition, `NashornScriptEngine` implements `Compilable`, so we can compile the JavaScript on policy create/update, rather than per evaluation.

pedroigor committed:
[KEYCLOAK-5056] - @NoCache to scope admin api

pedroigor committed:
JS policy should use ScriptingSPI

pedroigor reviewed (changes requested):
Thanks for these changes.

Regarding the ```ScriptingSPI```, here is a branch with some initial changes [1] to get it working with the ```JSPolicyProvider```.

I think you can adapt your code to also use the ```ScriptingSPI``` and maybe have there what you did around pre-compiled scripts.

What do you say ?

[1] https://github.com/pedroigor/keycloak/commit/0b5e6b0d497a596fa24e56f367037ed60a716b78

pedroigor commented:
@janslow , I've created https://issues.jboss.org/browse/KEYCLOAK-5072 as a umbrella for all changes we are discussing.

pedroigor commented on authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/js/JSPolicyProviderFactory.java:
```diff
@@ -19,9 +24,11 @@
-    private JSPolicyProvider provider = new JSPolicyProvider(() -> new ScriptEngineManager().getEngineByName(ENGINE));
+    private JSPolicyProvider provider = new JSPolicyProvider(this::getCompiledCode);
+    private Compilable engine;
+    private Map<String, CompiledScript> compiled = Collections.synchronizedMap(new HashMap<>());
```
We can probably use a simple LRU map to cache entries. Where the number of cached entries can be defined by users.

There is a simple LRU map implementation we are already using in other places, we can probably re-use the same code here.

janslow replied:
Could you point me at the LRU implementation? I couldn't find it.

janslow commented:
Sure, I'll rebase on your branch.

janslow committed:
Extract EvaluatebleScriptAdapter

Precursor for InvocableScriptAdapter, which compiles/evaluates a script without affecting the engine's bindings. This allows the same script to be compiled once and then evaluated multiple times (with the same ScriptEngine).

janslow committed:
Cache compiled scripts in JSPolicyProviderFactory

janslow commented:
I've reapplied my changes on top of your branch.

Originally, the `JSPolicyProvider` was evaluating the JavaScript twice, so I extracted a `EvaluatableScriptAdapter` (which can evaluate a script with some custom bindings), which is used by the JSPolicyProvider and DefaultScriptingProvider (to convert into an `InvocableScriptAdapter`).

I haven't solved the issue of tracking `ScriptModel`s in the `DefaultScriptingProvider`, as that may start to raise concurrency issues. Nashorn's `ScriptEngine` and `CompiledScript` are [thread-safe](https://stackoverflow.com/questions/30140103/should-i-use-a-separate-scriptengine-and-compiledscript-instances-per-each-threa/30159424#30159424), whilst `Bindings` isn't, so when the `ScriptEngine` is associated with a `Bindings`, it is no longer thread safe. This could be avoided by using [something like this](https://stackoverflow.com/a/36020936), but that would involve more Nashorn specific code.

pedroigor commented:
Regarding the issue you mentioned, I think we don't need to worry about it and just leave as it is. So we handle bindings accordingly within the new code you added to policy evaluation.

Will merge this PR today.

janslow commented:
That's great, thanks!
